### DISCUSSION TITLE
Improve Command Executor interface (Cmd)

### DISCUSSION THREAD

**Discussion by Tayipc38** [YimMenu/YimMenu#2860] [Ideas]
Currently the command executor remains somewhat raw and hard to use. It can benefit from QoL features such as predictive text and command history. This is entirely possible with the callback system ImGui has. My reason for starting a discussion first is to see whether such an improvement would have an audience.

Think of:
-Spawning cars offering you autocomplete and a list that best match your current input
-Player commands such as explode or multikick offering autocomplete on player names and also a similarity list.
-Maybe even very custom predictive text for specific commands such as 'joinsession' showing what session types can be put in.
-Many more like the ones above

I have already experimented with this, and can personally say using the menu through the cmd is much faster.
If you are in favor of this improvement. Show your support by commenting or upvoting this discussion.

**Comment by Yimura** [comment]
Completely agree, the initial command executor was a good starting point, however I’d like to add some additional functionality like array up and down for history support.

**Reply by Tayipc38** [reply]
What do you say of an extra override func on commands to allow for custom predictive text?

**Reply by Yimura** [reply]
That would be nice, the command executor asks for an autofill request on argument X in the command. After which the command gives a suggestion, however no clue how the actual UI will be implemented on that. Perhaps there's a text suggestion popup type thing already in ImGui. We should probably check the ImGui demo app for this.

**Reply by Tayipc38** [reply]
Currently, I simply show the most relevant suggestion right beneath the input. History is only shown when the input is empty, and you can scroll through the history using arrow keys. Pressing tab fills the input with the most relevant suggestion. History is limited to 10 commands and does not append if the command already is within that limit of 10. Ive refactored commands and suggestions to accept any argument that is close enough to a valid result. This fastens the process of using commands. For example this would work:
`tp yim day`. This would teleport you to me with my current commands setup. Both arguments would be supported with the full player list and the most relevant suggestion.
